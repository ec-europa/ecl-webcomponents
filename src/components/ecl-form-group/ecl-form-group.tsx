import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'ecl-form-group',
  styleUrls: {
    ec: './build/styles/ecl-form-group-ec.css',
    eu: './build/styles/ecl-form-group-eu.css',
  },
  shadow: false,
  scoped: true,
})

export class EclFormGroup {
  @Prop() styleClass: string = '';
  @Prop() theme: string = 'ec';
  @Prop() label: string;
  @Prop() invalid: boolean = false;
  @Prop() required: boolean = false;
  @Prop() requiredText: string = '*';
  @Prop() optionalText: string;
  @Prop() invalidText: string;
  @Prop() helperText: string;
  @Prop() helperId: string;
  @Prop() labelTag: string = 'label';
  @Prop() tag: string = 'div';
  @Prop() name: string;
  @Prop() invalidIcon: string = 'error';
  @Prop() labelClass: string;
  @Prop() ariaLabelOptional: string;
  @Prop() ariaLabelRequired: string;

  getClass(): string {
    return [
      `ecl-form-group`,
      this.styleClass
    ].join(' ');
  }

  render() {
    return (
      <this.tag
        class={this.getClass()}
        aria-describedby={this.helperId}
      >
      { this.label ?
        <this.labelTag
          class={`ecl-form-label ${this.invalid ? 'ecl-form-label--invalid' : ''} ${this.labelClass || ''}`}
          {...(this.ariaLabelRequired ? { 'aria-label': this.ariaLabelRequired } : {})}
        >
          {this.label}
      {this.required && this.requiredText ? (
        <span
          class="ecl-form-label__required" 
          {...(this.ariaLabelRequired ? { 'aria-label': this.ariaLabelRequired } : {})}
        >
          {this.requiredText}
        </span>
      ) : ''}
      { !this.required && this.optionalText ?
        <span
          class="ecl-form-label__optional"
          {...(this.ariaLabelOptional ? { 'aria-label': this.ariaLabelOptional } : {})}
        >
          {this.optionalText}</span> : ''
      }
        </this.labelTag> : ''
      }
      { this.helperText ? 
        <div
          class="ecl-help-block"
        >
          {this.helperText}
        </div> : ''
      }
      { this.invalid && this.invalidText ?
        <div class="ecl-feedback-message">
        { this.invalidIcon ? 
          <ecl-icon
            styleClass={`ecl-feedback-message__icon sc-ecl-form-group-${this.theme}`}
            icon={this.invalidIcon}
            size="m"
          ></ecl-icon> : '' 
        }
          {this.invalidText}
        </div> : '' 
      }
        <slot></slot>
      </this.tag>
    )
  }
}