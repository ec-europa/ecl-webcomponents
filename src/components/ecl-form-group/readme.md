# ecl-form-group



<!-- Auto Generated Below -->


## Properties

| Property            | Attribute             | Description | Type      | Default     |
| ------------------- | --------------------- | ----------- | --------- | ----------- |
| `ariaLabelOptional` | `aria-label-optional` |             | `string`  | `undefined` |
| `ariaLabelRequired` | `aria-label-required` |             | `string`  | `undefined` |
| `helperId`          | `helper-id`           |             | `string`  | `undefined` |
| `helperText`        | `helper-text`         |             | `string`  | `undefined` |
| `invalid`           | `invalid`             |             | `boolean` | `false`     |
| `invalidIcon`       | `invalid-icon`        |             | `string`  | `'error'`   |
| `invalidText`       | `invalid-text`        |             | `string`  | `undefined` |
| `label`             | `label`               |             | `string`  | `undefined` |
| `labelClass`        | `label-class`         |             | `string`  | `undefined` |
| `labelTag`          | `label-tag`           |             | `string`  | `'label'`   |
| `name`              | `name`                |             | `string`  | `undefined` |
| `optionalText`      | `optional-text`       |             | `string`  | `undefined` |
| `required`          | `required`            |             | `boolean` | `false`     |
| `requiredText`      | `required-text`       |             | `string`  | `'*'`       |
| `styleClass`        | `style-class`         |             | `string`  | `''`        |
| `tag`               | `tag`                 |             | `string`  | `'div'`     |
| `theme`             | `theme`               |             | `string`  | `'ec'`      |


## Dependencies

### Used by

 - [ecl-search-form](../ecl-search-form)

### Depends on

- [ecl-icon](../ecl-icon)

### Graph
```mermaid
graph TD;
  ecl-form-group --> ecl-icon
  ecl-search-form --> ecl-form-group
  style ecl-form-group fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
