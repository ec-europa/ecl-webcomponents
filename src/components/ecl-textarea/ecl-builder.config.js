const path = require('path');
const pkg = require('./package.json');

const nodeModules = path.resolve(__dirname, '../../../node_modules');
const outputFolder = './build';

// SCSS includePaths
const includePaths = [nodeModules];

const banner = `${pkg.name} - ${pkg.version} Built on ${new Date().toISOString()}`;

module.exports = {
  styles: [
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-textarea-ec.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-textarea-ec.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-textarea-eu.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-textarea-eu.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
  ],
};
