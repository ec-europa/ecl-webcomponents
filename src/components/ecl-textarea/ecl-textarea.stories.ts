const getArgs = () => {
  return {
    required: true,
    disabled: false,
    invalid: false,
    width: 'm',
    label: 'Label',
    textareaId: 'textarea-id',
    helperText: `This is the input helper text.`,
    requiredText: '*',
    invalidText: 'This is the error message',
    invalidIcon: 'error',
    optionalText: '(optional)',
  };
};

const getArgTypes = () => {
  return {
    required: {
      type: { name: 'boolean' },
      description: 'Required input',
      table: {
        category: 'States',
      },
    },
    disabled: {
      type: { name: 'boolean' },
      description: 'Disabled input',
      table: {
        category: 'States',
      },
    },
    invalid: {
      type: { name: 'boolean' },
      description: 'Invalid input',
      table: {
        category: 'States',
      },
    },
    label: {
      type: { name: 'string' },
      description: 'Legend of the form group',
      table: {
        category: 'Content',
      },
    },
    requiredText: {
      name: 'required-text',
      type: { name: 'string' },
      description: 'Text to show when required',
      table: {
        category: 'Content',
      },
    },
    optionalText: {
      name: 'optional-text',
      type: { name: 'string' },
      description: 'Text to show when optional',
      table: {
        category: 'Content',
      },
    },
    helperText: {
      name: 'helper-text',
      type: { name: 'string' },
      description: 'Helper text',
      table: {
        category: 'Content',
      },
    },
    invalidText: {
      name: 'invalid-text',
      type: { name: 'string' },
      description: 'Error message',
      table: {
        category: 'Content',
      },
    },
    invalidIcon: {
      name: 'invalid-icon',
      type: { name: 'string' },
      description: 'Error icon',
      table: {
        category: 'Content',
      },
    },
    width: {
      type: { name: 'select' },
      description: 'Size of the textarea',
      options: ['s', 'm', 'l'],
      table: {
        category: 'Styles',
      },
    },
    textareaId: {
      name: 'textarea-id',
      type: { name: 'string' },
      description: 'Id of the textarea',
      table: {
        category: 'Content',
      },
    },
  };
};

export default {
  title: 'Components/form/textarea',
};

const Template = (args) => 
  `<ecl-form-group
    theme="${args.theme}"
    helper-text="${args.helperText}"
    helper-id="helper-id"
    invalid=${args.invalid}
    invalid-text="${args.invalidText}"
    invalid-icon="${args.invalidIcon}"
    optional-text="${args.optionalText}"
    label="${args.label}"
    required=${args.required}
    disabled="${args.disabled}"
  >
    <ecl-textarea
      theme="${args.theme}"
      required=${args.required}
      invalid=${args.invalid}
      disabled=${args.disabled}
      name="ecl-textarea"
      width="${args.width}"
      textarea-id=${args.textareaId}
    >
    </ecl-textarea>
  </ecl-form-group>`;

export const Textarea = Template.bind({});
Textarea.storyName = 'default';
Textarea.args = getArgs();
Textarea.argTypes = getArgTypes();