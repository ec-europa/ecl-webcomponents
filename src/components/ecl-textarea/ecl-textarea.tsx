import { Component, h, Prop, Element, Event, EventEmitter} from '@stencil/core';

@Component({
  tag: 'ecl-textarea',
  styleUrls: {
    ec: './build/styles/ecl-textarea-ec.css',
    eu: './build/styles/ecl-textarea-eu.css',
  },
  shadow: false,
  scoped: true,
})
export class EclTextarea{
  @Element() el: HTMLElement;
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() disabled: boolean = false;
  @Prop() required: boolean = false;
  @Prop() invalid: boolean = false;
  @Prop() width: string = 'm';
  @Prop() textareaId: string;
  @Prop() name: string;
  @Prop() rows: number = 4;
  @Prop() inputId: string;
  @Prop() placeholder: string;
  @Prop() defaultValue: string;
  @Prop() hasChanged: boolean = false;
  @Prop() isFocused: boolean = false;
  @Event() inputFocus: EventEmitter<FocusEvent>;
  @Event() inputBlur: EventEmitter<FocusEvent>;
  @Event() inputChange: EventEmitter;

  componentDidRender() {
    if (this.textareaId) {
      const group = this.el.closest('.ecl-form-group');
      if (group) {
        const label =  group.querySelector('.ecl-form-label');
        if (label) {
          label.setAttribute('for', this.textareaId);
          label.setAttribute('id', `${this.textareaId}-label`);
        }
        const helper = group.querySelector('.ecl-help-block');
        if (helper) {
          helper.setAttribute('id', `${this.textareaId}-helper`);
        }
      }
    }
  }

  getClass(): string {
    const styleClasses = [
      `ecl-text-area`,
      `ecl-text-area--${this.width}`,
      this.styleClass
    ];

    if (this.disabled) {
      styleClasses.push('ecl-text-area--disabled');
    }
    if (this.invalid) {
      styleClasses.push('ecl-text-area--invalid');
    }

    return styleClasses.join(' ');
  }

  handleFocus(event) {
    this.inputFocus.emit(event);
    this.isFocused = true;
  }

  handleChange(event) {
    this.inputChange.emit(event);
    this.hasChanged = true;
  }

  handleBlur(event) {
    this.inputBlur.emit(event);
    this.isFocused = false;
  }

  render() {
    let attributes = {
      class: this.getClass(),
      id: this.textareaId,
      name: this.name,
      required: this.required,
      disabled: this.disabled,
      placeholder: this.placeholder,
      rows: this.rows,
      value: this.defaultValue,
    };

    return (
      <textarea
        {...attributes}
        onInput={ev => this.handleChange(ev)}
        onFocus={ev => this.handleFocus(ev)}
        onBlur={ev => this.handleBlur(ev)}
        onChange={ev => this.handleChange(ev)}
      >
          <slot></slot>
      </textarea>
    );
  }
}