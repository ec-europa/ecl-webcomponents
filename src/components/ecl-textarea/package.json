{
  "name": "@ecl/web-components-text-area",
  "author": "European Commission",
  "license": "EUPL-1.2",
  "version": "0.2.0",
  "description": "ECL Textarea",
  "main": "ecl-textarea.tsx",
  "private": true,
  "scripts": {
    "build:styles": "cross-env NODE_ENV=production ecl-builder styles",
    "build": "run-s clean build:*",
    "clean": "rimraf build",
    "prepublish": "npm run build"
  },
  "devDependencies": {
    "@ecl/builder": "3.7.1",
    "@ecl/vanilla-component-text-area": "3.7.1",
    "cross-env": "7.0.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/planctus/ecl-webcomponents.git"
  },
  "bugs": {
    "url": "https://github.com/planctus/ecl-webcomponents/issues"
  },
  "homepage": "https://github.com/planctus/ecl-webcomponents",
  "keywords": [
    "ecl",
    "europa-component-library",
    "design-system",
    "web-components"
  ]
}
