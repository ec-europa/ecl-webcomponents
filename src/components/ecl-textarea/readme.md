# ecl-textarea



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute     | Description | Type      | Default     |
| ------------- | ------------- | ----------- | --------- | ----------- |
| `disabled`    | `disabled`    |             | `boolean` | `false`     |
| `hasChanged`  | `has-changed` |             | `boolean` | `false`     |
| `invalid`     | `invalid`     |             | `boolean` | `false`     |
| `isFocused`   | `is-focused`  |             | `boolean` | `false`     |
| `name`        | `name`        |             | `string`  | `undefined` |
| `placeholder` | `placeholder` |             | `string`  | `undefined` |
| `required`    | `required`    |             | `boolean` | `false`     |
| `rows`        | `rows`        |             | `number`  | `4`         |
| `styleClass`  | `style-class` |             | `string`  | `undefined` |
| `textareaId`  | `textarea-id` |             | `string`  | `undefined` |
| `theme`       | `theme`       |             | `string`  | `'ec'`      |
| `width`       | `width`       |             | `string`  | `'m'`       |


## Events

| Event         | Description | Type                      |
| ------------- | ----------- | ------------------------- |
| `inputBlur`   |             | `CustomEvent<FocusEvent>` |
| `inputChange` |             | `CustomEvent<any>`        |
| `inputFocus`  |             | `CustomEvent<FocusEvent>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
