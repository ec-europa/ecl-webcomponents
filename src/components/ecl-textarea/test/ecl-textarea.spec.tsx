import { newSpecPage } from '@stencil/core/testing';
import { EclTextarea } from '../ecl-textarea';
import { EclFormGroup } from '../../ecl-form-group/ecl-form-group';

describe('ecl-textarea', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [EclTextarea, EclFormGroup],
      html:  `<ecl-form-group
        helper-text="Helper text"
        helper-id="helper-id"
        label="Select input field"
      >
        <ecl-textarea
          required
          name="ecl-textarea"
          width="l"
        >
        </ecl-textarea>
      </ecl-form-group>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
