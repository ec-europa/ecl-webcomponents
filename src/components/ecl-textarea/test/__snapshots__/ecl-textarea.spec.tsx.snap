// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-textarea renders 1`] = `
<ecl-form-group helper-id="helper-id" helper-text="Helper text" label="Select input field">
  <!---->
  <div aria-describedby="helper-id" class="ecl-form-group">
    <label class="ecl-form-label">
      Select input field
    </label>
    <div class="ecl-help-block">
      Helper text
    </div>
    <ecl-textarea name="ecl-textarea" required="" width="l">
      <!----><textarea class="ecl-text-area ecl-text-area--l" name="ecl-textarea" required="" rows="4">
        </textarea>
    </ecl-textarea>
  </div>
</ecl-form-group>
`;
